### PR TITLE
Add local search sampler

### PR DESCRIPTION
**Pull request by hyeok9855**
This is a PR for adding `LocalSearchSampler`.
The local search is based on the work [[1]](https://arxiv.org/abs/2202.01361) and [[2]](https://arxiv.org/abs/2310.02710).

Test in hypergrid env:
```bash
python tutorials/examples/train_hypergrid_simple_ls.py
```

### CODE REVIEW

**Comment by josephdviviano** [comment]
@hyeok9855 can you fix the merge conflicts?

**Comment by josephdviviano** [comment]
I noticed you are using `force-push` -- be careful with this, it can put the code in a state hard to resolve with the rest of the work. 

https://www.gitkraken.com/learn/git/problems/git-push-force#:~:text=The%20Risks%20of%20Git%20Push%20Force&text=Because%20you%20have%20failed%20to,deleting%20your%20team%20member's%20work.

**Comment by hyeok9855** [comment]
I fixed an issue in the backward mask!

Is there anything necessary to do next?

File: src/gfn/samplers.py [hunk 1]
```diff
@@ -264,3 +263,324 @@ def sample_trajectories(
...
+            # TODO: Implement Metropolis-Hastings acceptance criterion.
+            # p(x->s'->x') = p_B(x->s')p_F(s'->x')
+            # p(x'->s'->x) = p_B(x'->s')p_F(s'->x)
+            # The acceptance ratio is
+            # min(1, R(x')p(x->s'->x') / R(x)p(x'->s'-> x))
+            # Note that
+            # p(x->s'->x') / p(x'->s'-> x))
+            # = p_B(x->s')p_F(s'->x') / p_B(x'->s')p_F(s'->x)
+            # = p_B(x->s'->s0)p_F(s0->s'->x') / p_B(x'->s'->s0)p_F(s0->s'->x)
+            # = p_B(tau|x)p_F(tau') / p_B(tau'|x')p_F(tau)
```
Comment (hyeok9855) [inline src/gfn/samplers.py:497]: This is how Metropolis-Hastings works.
In the last line, we need both log_pfs and log_pbs of each trajectory.

File: src/gfn/samplers.py [hunk 2]
```diff
@@ -264,3 +263,324 @@ def sample_trajectories(
...
+            conditioning=conditioning,
+            save_estimator_outputs=save_estimator_outputs,
+            save_logprobs=save_logprobs,
+            **policy_kwargs,
+        )
+        # Calculate the forward probability if needed (metropolis-hastings).
+        ### COPIED FROM `TrajectoryBasedGFlowNet.get_pfs_and_pbs` ###
+        if use_metropolis_hastings:
+            ### FIXME: I realize that the trajectory needs to be reverted to get the forward probability.
+            ### TODO: Resolve the issue first https://github.com/GFNOrg/torchgfn/issues/109
```
Comment (hyeok9855) [inline src/gfn/samplers.py:328]: The backward trajectory `(e.g., [2, 1] -> [1, 1] -> [1, 0] -> [0, 0] -> [0, 0] (dummy) ...)` is quite different from the forward trajectory `(e.g., [0, 0] -> [1, 0] -> [1, 1] -> [2, 1] -> [-1, -1] (terminal) -> [-1, -1] -> ...)`.

Thus, we cannot compute pf for bwd. traj. in the same way as in `TrajectoryBasedGFlowNet` which calculates pf for fwd. traj.

I believe there must be some other way to resolve this, but I guess it would be the simplest if we can revert the traj. directions, check this related issue #109.

File: src/gfn/samplers.py [hunk 3]
```diff
@@ -264,3 +263,324 @@ def sample_trajectories(
...
+                valid_actions.tensor
+            )  # Using the actions sampled off-policy.
+            log_pf_backward_trajectories = torch.full_like(
+                backward_trajectories.actions.tensor[..., 0],
+                fill_value=0.0,
+                dtype=torch.float,
+            )
+            log_pf_backward_trajectories[
+                ~backward_trajectories.actions.is_dummy
+            ] = valid_log_pf_actions
```
Comment (hyeok9855) [inline src/gfn/samplers.py:364]: This part is copied from `TrajectoryBasedGFlowNet.get_pfs_and_pbs`.

I couldn't find any other way to calculate the `log_pfs` or `log_pbs`, but this is too ugly and too much redundancies.

We need more modular functionality to minimize the redundancy.

File: src/gfn/samplers.py [hunk 4]
```diff
@@ -264,3 +263,324 @@ def sample_trajectories(
...
+                fill_value=0.0,
+                dtype=torch.float,
+            )
+            log_pb_new_trajectories_slice = torch.full_like(
+                valid_actions.tensor[..., 0], fill_value=0.0, dtype=torch.float
+            )
+            log_pb_new_trajectories_slice[~valid_actions.is_exit] = valid_log_pb_actions
+            log_pb_new_trajectories[
+                ~new_trajectories.actions.is_dummy
+            ] = log_pb_new_trajectories_slice
```
Comment (hyeok9855) [inline src/gfn/samplers.py:486]: DITTO

File: src/gfn/samplers.py [hunk 5]
```diff
@@ -264,3 +268,275 @@ def sample_trajectories(
...
+            env,
+            states=trajectories.last_states,
+            conditioning=conditioning,
+            save_estimator_outputs=save_estimator_outputs,
+            save_logprobs=save_logprobs,
+            **policy_kwargs,
+        )
+
+        # Calculate the forward probability if needed (Metropolis-Hastings).
+        prev_trajectories = Trajectories.reverse_backward_trajectories(
```
Comment (josephdviviano) [inline src/gfn/samplers.py:335]: maybe for clarity `prev_forward_trajectories` or `reversed_bakward_trajectories`, since you also operate on `trajectories` below.

File: src/gfn/samplers.py [hunk 6]
```diff
@@ -264,3 +268,275 @@ def sample_trajectories(
...
+            backward_trajectories
+        )
+        prev_trajectories_log_rewards = trajectories.log_rewards
```
Comment (josephdviviano) [inline src/gfn/samplers.py:338]: should this be `prev_trajectories`? I actually think it does not matter since you're just looking at the reward at the end of the trajectory, but a comment here would be clarifying b/c above `prev_trajectories` refers to the reverse of a trajectory sampled from pf and here you're grabbing log_rewards directly from the forward trajectories.

File: src/gfn/samplers.py [hunk 7]
```diff
@@ -264,3 +268,275 @@ def sample_trajectories(
...
+    by Zhang et al. 2022 (https://arxiv.org/abs/2202.01361) for negative sampling
+    and further explored its effectiveness in various applications by Kim et al. 2023
+    (https://arxiv.org/abs/2310.02710).
+
+    Attributes:
+        estimator: the submitted PolicyEstimator for the forward pass.
+        pb_estimator: the PolicyEstimator for the backward pass.
+    """
+
+    def __init__(self, estimator: GFNModule, pb_estimator: GFNModule):
```
Comment (josephdviviano) [inline src/gfn/samplers.py:285]: let's rename `estimator` to `pf_estimator`

File: src/gfn/samplers.py [hunk 8]
```diff
@@ -264,3 +268,275 @@ def sample_trajectories(
...
+            )
+            log_pb_new_trajectories = torch.full((max_traj_len - 1, bs), 0.0).to(
+                device=device, dtype=torch.float
+            )
+
+        for i in range(bs):  # FIXME: Can we vectorize this?
+            n_back = backward_trajectories.when_is_done[i] - K[i]
+
+            # Sanity check
+            assert (
```
Comment (josephdviviano) [inline src/gfn/samplers.py:394]: It would be good I think to move this into a test, to prevent the assertion being run so often in production.

File: src/gfn/samplers.py [hunk 9]
```diff
@@ -264,3 +268,275 @@ def sample_trajectories(
...
+                trajectories=prev_trajectories,
+            )
+            log_pb_recon_trajectories = get_trajectory_pbs(
+                pb=self.backward_sampler.estimator, trajectories=recon_trajectories
+            )
+            log_pb_new_trajectories = torch.full((max_traj_len - 1, bs), 0.0).to(
+                device=device, dtype=torch.float
+            )
+
+        for i in range(bs):  # FIXME: Can we vectorize this?
```
Comment (josephdviviano) [inline src/gfn/samplers.py:390]: I think this should be vectorized -- if you need help, no problem, let's schedule a pair programming session (maybe in the evening EST instead of my early morning ;) ).

**Review by josephdviviano — CHANGES_REQUESTED** [review]
Hey @hyeok9855 - please see my comments. This is a really awesome PR, I like the changes you made to `trajectories` and with some tweaks the local search sampler looks very clean. 
 
Let me know if you want to schedule a pair programming session.

File: src/gfn/containers/trajectories.py [hunk 10]
```diff
@@ -428,6 +427,68 @@ def to_non_initial_intermediary_and_terminating_states(
             conditioning,
         )
 
+    @staticmethod
+    def reverse_backward_trajectories(trajectories: Trajectories) -> Trajectories:
+        """Reverses a backward trajectory"""
+        # FIXME: This method is not compatible with continuous GFN.
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/containers/trajectories.py:433]: I'm not sure either... This was from [here](https://github.com/GFNOrg/torchgfn/issues/109).

File: src/gfn/containers/trajectories.py [hunk 11]
```diff
@@ -428,6 +427,68 @@ def to_non_initial_intermediary_and_terminating_states(
...
+            )
+
+        new_when_is_done = trajectories.when_is_done + 1
+        new_states = trajectories.env.sf.repeat(
+            new_when_is_done.max() + 1, len(trajectories), 1
+        )
+
+        # FIXME: Can we vectorize this?
+        # FIXME: Also, loop over batch or sequence?
+        for i in range(len(trajectories)):
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/containers/trajectories.py:461]: Of course, yes. I'll check and let you know if I need your help.

File: src/gfn/env.py [hunk 12]
```diff
@@ -61,6 +61,8 @@ def __init__(
         self.dummy_action = dummy_action
         self.exit_action = exit_action
 
+        # Warning: don't use self.States or self.Actions to initialize an instance of the class.
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/env.py:64]: Maybe us?? Regarding this, what about making them into private variables (e.g., `self.__States` and `self.__Actions`)??

File: src/gfn/samplers.py [hunk 13]
```diff
@@ -242,9 +246,9 @@ def sample_trajectories(
             trajectories_states.append(deepcopy(states))
 
         trajectories_states = stack_states(trajectories_states)
-        trajectories_actions = env.Actions.stack(trajectories_actions)
+        trajectories_actions = env.Actions.stack(trajectories_actions)[1:, :]
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/samplers.py:249]: Oh, sure. It's just to ignore the dummy actions (check line 158 of this file).

File: src/gfn/samplers.py (hunk 9 above)
Comment (hyeok9855) [reply src/gfn/samplers.py:390]: Let me check first! I will let you know :)

File: src/gfn/samplers.py (hunk 8 above)
Comment (hyeok9855) [reply src/gfn/samplers.py:394]: Is it possible to test with the local variables of a function?

File: src/gfn/samplers.py (hunk 6 above)
Comment (hyeok9855) [reply src/gfn/samplers.py:338]: Why this is `prev_trajectories` -> check like 450-456. I wanted that part to be `new...` +/- `prev...`.
To alleviate the confusion, we can simply change this line to 
```
prev_trajectories_log_rewards = prev_trajectories.log_rewards
```

File: src/gfn/samplers.py (hunk 5 above)
Comment (hyeok9855) [reply src/gfn/samplers.py:335]: I wanted to make this correspond to the `new_trajectories.` (Please check my comment below.)

I think adding a short explanation of why this is called `prev_trajectories` would be fine, e.g., 
```
        # By reversing the backward trajectories, obtain the forward trajectories.
        # This is called `prev_trajectories` since they are the trajectories before
        # the local search. The `new_trajectories` will be obtained by performing local
        # search on them.
        prev_trajectories = Trajectories.reverse_backward_trajectories(
            backward_trajectories
        )
```

What do you think about this??

**Review by younik — COMMENTED** [review]
Good job, the code is very well written :)

File: src/gfn/samplers.py [hunk 14]
```diff
@@ -169,10 +175,8 @@ def sample_trajectories(
         all_estimator_outputs = []
 
         while not all(dones):
-            actions = env.actions_from_batch_shape((n_trajectories,))  # Dummy actions.
-            log_probs = torch.full(
-                (n_trajectories,), fill_value=0, dtype=torch.float, device=device
-            )
+            actions = deepcopy(dummy_actions)
+            log_probs = dummy_logprobs.clone()
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/samplers.py:179]: If `states` is given, we start from there.
In a very edge case, `all(dones)` is True at the beginning (i.e., all `states` are already terminal!).
Previously, `trajectories_actions` and `trajectories_dones` were empty lists in this edge case, and this raises some error with `torch.stack`.

File: src/gfn/samplers.py [hunk 15]
```diff
@@ -264,3 +272,298 @@ def sample_trajectories(
...
+            save_estimator_outputs,
+            save_logprobs or use_metropolis_hastings,
+            **policy_kwargs,
+        )
+
+        if n is None:
+            n = trajectories.n_trajectories
+
+        search_indices = torch.arange(n, device=trajectories.states.device)
+        for it in range(n_local_search_loops - 1):
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/samplers.py:549]: Very nice

File: src/gfn/samplers.py [hunk 16]
```diff
@@ -242,9 +246,13 @@ def sample_trajectories(
             trajectories_states.append(deepcopy(states))
 
         trajectories_states = stack_states(trajectories_states)
-        trajectories_actions = env.Actions.stack(trajectories_actions)
+        trajectories_actions = env.Actions.stack(trajectories_actions)[
+            1:, ...  # Drop dummy action
+        ]
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/samplers.py:251]: Can I ask why do you think it is better?
I think `...` could indicate that this is not a 1d Tensor.

File: src/gfn/samplers.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (younik) [reply src/gfn/samplers.py:179]: Oh, I see, thanks!

File: src/gfn/samplers.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (younik) [reply src/gfn/samplers.py:251]: It is very minor, but it allocates a bigger tensor than the one you need. When you call `[1:]` it creates a new view, but the full tensor remains in memory

File: src/gfn/containers/trajectories.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/containers/trajectories.py:433]: One guess is this:

In line 436-443:
```
        new_actions = torch.full(
            (
                trajectories.max_length + 1,
                len(trajectories),
                *trajectories.actions.action_shape,
            ),
            -1,
        )
```
Also, in line 461-463:
```
            new_actions[trajectories.when_is_done[i], i] = (
                trajectories.env.n_actions - 1
            )
```

These assume that the action is an integer, which is not true for continuous case, right?

File: src/gfn/samplers.py [hunk 17]
```diff
@@ -264,3 +272,298 @@ def sample_trajectories(
...
+                **policy_kwargs,
+            )
+            trajectories.extend(ls_trajectories)
+
+            last_indices = torch.arange(
+                n * (it + 1), n * (it + 2), device=trajectories.states.device
+            )
+            search_indices[is_updated] = last_indices[is_updated]
+
+        return trajectories
```
Comment (saleml) [inline src/gfn/samplers.py:569]: this is great !

File: src/gfn/utils/prob_calculations.py [hunk 18]
```diff
@@ -103,23 +104,34 @@ def get_trajectory_pfs(
             valid_actions.tensor
         )  # Using the actions sampled off-policy.
 
-        log_pf_trajectories = torch.full_like(
-            trajectories.actions.tensor[..., 0],
```
Comment (saleml) [inline src/gfn/utils/prob_calculations.py:107]: what's the rationale behind removing this?

File: src/gfn/utils/prob_calculations.py [hunk 19]
```diff
@@ -145,13 +160,13 @@ def get_trajectory_pbs(
         valid_states, estimator_outputs
     ).log_prob(valid_actions.tensor)
 
-    log_pb_trajectories = torch.full_like(
-        trajectories.actions.tensor[..., 0],
-        fill_value=fill_value,
```
Comment (saleml) [inline src/gfn/utils/prob_calculations.py:150]: same question here

**Review by saleml — COMMENTED** [review]
great work

File: src/gfn/containers/trajectories.py [hunk 20]
```diff
@@ -451,26 +451,94 @@ def reverse_backward_trajectories(trajectories: Trajectories) -> Trajectories:
...
+
+        # Transpose back
+        new_actions = new_actions.transpose(
+            0, 1
+        )  # shape (max_len + 1, n_trajectories, *action_dim)
+        new_states = new_states.transpose(
+            0, 1
+        )  # shape (max_len + 2, n_trajectories, *state_dim)
+
+        # TODO: Add below into the test suite to ensure correctness
```
Comment (saleml) [inline src/gfn/containers/trajectories.py:522]: thank you for handling the vectorization. did you test this?

File: src/gfn/utils/prob_calculations.py (hunk 18 above)
Comment (hyeok9855) [reply src/gfn/utils/prob_calculations.py:107]: This was just moved to line 78 to address the edge case in line 84!

File: src/gfn/utils/prob_calculations.py (hunk 19 above)
Comment (hyeok9855) [reply src/gfn/utils/prob_calculations.py:150]: Same here :)

File: src/gfn/containers/trajectories.py (hunk 20 above)
Comment (hyeok9855) [reply src/gfn/containers/trajectories.py:522]: Yes, I test this by uncommenting it.
@josephdviviano, Could you give any advice on how to design a test to check whether the vectorization works appropriately?

Comment (saleml) [reply src/gfn/containers/trajectories.py:522]: What about copying the code pre-vecotrization in a test file and compare the outputs of your new function and that code on a few hypergrid + other env trajectories ?

File: testing/test_samplers_and_trajectories.py [hunk 21]
```diff
@@ -214,3 +214,86 @@ def test_replay_buffer(
         replay_buffer.add(training_objects)
     except Exception as e:
         raise ValueError(f"Error while testing {env_name}") from e
+
+
+@pytest.mark.parametrize("env_name", ["HyperGrid", "DiscreteEBM"])
+def test_reverse_backward_trajectories(env_name: str):
```
Comment (saleml) [inline testing/test_samplers_and_trajectories.py:220]: great

**Comment by hyeok9855** [comment]
FYI: In terms of L1 distance (empirical distribution vs. true distribution, as did in [original GFN paper](https://arxiv.org/abs/2106.04399)), **both LS-GFN (TB) w/ and w/o Metropolis-Hastings correction outperform vanilla TB** in 16x16x16x16 HyperGrid (using the default hyperparameters).

- TB: $2.571 \times 10^{-5}$
- TB + LS: $2.513 \times 10^{-5}$
- TB + LS + MH: $2.507 \times 10^{-5}$

**Comment by saleml** [comment]
LGTM.
@josephdviviano , please merge if you're satisfied with the changed.
